File: curso-dev-dojo-spring-essentials/src/main/java/com/invillia/CursoSpringEssentialsDevDojo/config/SecurityConfig.java
package com.invillia.CursoSpringEssentialsDevDojo.config;

import com.invillia.CursoSpringEssentialsDevDojo.service.CustomUserDetailService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.config.annotation.method.configuration.EnableGlobalMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;

@EnableWebSecurity
@EnableGlobalMethodSecurity(prePostEnabled = true)
public class SecurityConfig extends WebSecurityConfigurerAdapter {

    private final CustomUserDetailService customUserDetailService;

    @Autowired
    public SecurityConfig(CustomUserDetailService customUserDetailService) {
        this.customUserDetailService = customUserDetailService;
    }


    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http.authorizeRequests()
                //.anyRequest().authenticated()
                .antMatchers("/*/protected/**").hasRole("USER")// * -> alguma coisa antes de protected/que tenha protected/** -> tudo que vier depois
                .antMatchers("/*/admin/**").hasRole("ADMIN")
                .and()
                .httpBasic()
                .and()
                .csrf().disable();
    }

    //CONFIGURAÇÃO COM USUÁRIO NA MEMÓRIA
    @Autowired
    public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception{
        auth.inMemoryAuthentication()
                .withUser("igor").password("{noop}devdojo").roles("USER")
                .and()
                .withUser("admin").password("{noop}devdojo").roles("USER", "ADMIN");
    }

//    @Autowired
//    public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception{
//        auth.userDetailsService(customUserDetailService).passwordEncoder(new BCryptPasswordEncoder());
//    }


}

File: ponto-eletronico-kotlin-udemy/src/main/kotlin/com/curso/udemy/pontoeletronicokotlin/security/SecurityConfig.kt
package com.curso.udemy.pontoeletronicokotlin.security

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.authentication.dao.DaoAuthenticationProvider
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.config.annotation.method.configuration.EnableGlobalMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder


/**
 * @Configuration -> Mostrando pro spring que é uma classe de configuração
 * @EnableWebSecurity -> Habilitando o web security na aplicação
 * EnableGlobalMethodSecurity -> Validar o metodo especifico por perfil
 *
 */

@Configuration
@EnableWebSecurity
@EnableGlobalMethodSecurity(prePostEnabled = true)
class SecurityConfig(val funcionarioDetailService: FuncionarioDetailService) : WebSecurityConfigurerAdapter() {

    override fun configure(auth: AuthenticationManagerBuilder?) {
        auth?.authenticationProvider(authenticationProvider())
    }

    override fun configure(http: HttpSecurity?) {
        //Regras de segurança
        http?.
            authorizeRequests()?. // Autenticação nas requisições
            antMatchers("/api/cadastrar-pj", "/api/cadastrar-pf")?. // Fazer com que essas rotas sejam permitidas
            permitAll()?. // Permitindo rotas informadas no comando anterior
            anyRequest()?. // Autenticar todas requisições
            authenticated()?.and()?. // Garantir todas requisições autenticadas e
            httpBasic()?.and()?. // Será utilizada autenticação basica e
            sessionManagement()?. // Vou usar gerenciamento de sessão
            sessionCreationPolicy(SessionCreationPolicy.STATELESS)?.and()?. // Valida durante a requisição(desde o inicio da chamada até o retorno e)
            csrf()?.disable() // Desabilitando proteção de segurança
    }

    @Bean
    fun authenticationProvider(): DaoAuthenticationProvider {
        val authProvider = DaoAuthenticationProvider()
        authProvider.setUserDetailsService(funcionarioDetailService)
        authProvider.setPasswordEncoder(encoder())
        return authProvider
    }

    @Bean
    fun encoder(): PasswordEncoder = BCryptPasswordEncoder()
}